network/src/gossip/message_routing.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Data structures and synchronous logic for ICMP message gossip.
+
+use sr_primitives::traits::{ProvideRuntimeApi, BlakeTwo256, Hash as HashT};
+use sr_primitives::generic::BlockId;
+use polkadot_primitives::{Block, Hash};
+use polkadot_primitives::parachain::ParachainHost;
+use std::collections::{HashMap, HashSet};
+use super::{MAX_CHAIN_HEADS, LeavesVec};
+
+/// Construct a topic for a message queue root deterministically.
+pub fn queue_topic(queue_root: Hash) -> Hash {
+	let mut v = queue_root.as_ref().to_vec();
+	v.extend(b"message_queue");
+
+	BlakeTwo256::hash(&v[..])
+}
+
+/// A type-safe handle to a leaf.
+pub trait Leaf {
+	fn hash(&self) -> &Hash;
+}
+
+impl Leaf for Hash {
+	fn hash(&self) -> &Hash {
+		self
+	}
+}
+
+/// Context to the underlying polkadot chain.
+pub trait ChainContext {
+	/// The leaf type that this chain context is associated with.
+	type Leaf: Leaf;
+	/// An error that can occur when querying the chain.
+	type Error;
+
+	/// Provide a closure which is invoked for every unrouted queue hash at a given leaf.
+	fn leaf_unrouted_roots<F: FnMut(&Hash)>(&self, leaf: &Self::Leaf, with_queue_root: F)
+		-> Result<(), Self::Error>;
+}
+
+impl<P: ProvideRuntimeApi> ChainContext for P where P::Api: ParachainHost<Block> {
+	type Leaf = Hash;
+	type Error = substrate_client::error::Error;
+
+	fn leaf_unrouted_roots<F: FnMut(&Hash)>(&self, &leaf: &Self::Leaf, mut with_queue_root: F)
+		-> Result<(), Self::Error>
+	{
+		let api = self.runtime_api();
+
+		let leaf_id = BlockId::Hash(leaf);
+		let active_parachains = api.active_parachains(&leaf_id)?;
+
+		for para_id in active_parachains {
+			if let Some(ingress) = api.ingress(&leaf_id, para_id)? {
+				for (_height, _from, queue_root) in ingress.iter() {
+					with_queue_root(queue_root);
+				}
+			}
+		}
+
+		Ok(())
+	}
+}
+
+/// A view of which queue roots are current for a given set of leaves.
+pub struct View {
+	leaves: LeavesVec,
+	leaf_topics: HashMap<Hash, HashSet<Hash>>, // leaf_hash -> { topics }
```

Comment:
Could we declare and then import as necessary some type aliases like the following to avoid having to add comments to describe the implementation:
```
pub type LeafHash = T::Hash;
pub type LeafTopics = HashSet<T::Hash>;
```
```suggestion
	leaf_topics: HashMap<LeafHash, LeafTopics>,
```

Reply:
same reply as above